### ISSUE TITLE
C# highlighter breaks on generic base classes 

### ISSUE THREAD

**Issue by damirarh** [isagalaev/highlight.js#684]
The following snippet doesn't get highlighted:

    public class IncrementalLoadingBehavior : Behavior<LongListSelector>
    { }

If the base class isn't generic, it works as expected:

    public class IncrementalLoadingBehavior : Behavior
    { }

**Comment by citizenmatt** [comment]
Is this endless loop with the PR from #683 applied?

**Comment by citizenmatt** [comment]
Hmm. That code snippet works fine in the developer tool, but I've got a snippet that also causes problems. However, it's not an infinite loop, it's just MASSIVELY slow. Here's the snippet:

```cs
var cleanupProfile = BulkCodeCleanupActionBuilderBase.CreateProfile(profile =>
  {
    profile.SetSetting(ReplaceByVar.USE_VAR_DESCRIPTOR
  }, new ReplaceByVar.Options()
  {
    BehavourStyle = ReplaceByVar.BehavourStyle.CAN_CHANGE_TO_EXPLICIT,
    ForeachVariableStyle = ReplaceByVar.ForeachVariableStyle.ALWAYS_EXPLICIT,
    LocalVariableStyle = ReplaceByVar.LocalVariableStyle.ALWAYS_EXPLICIT
  }));

var projectFile = myProvider.SourceFile.ToProjectFile();
Assertion.AssertNotNull(projectFile, "projectFile must not be null");

var builder = BulkCodeCleanupContextActionBuilder.CreateByPsiLanguage<CSharpLanguage>(cleanupProfile,
  "Use explicit type everywhere", projectFile.GetSolution(), this);
return builder.CreateBulkActions(projectFile, IntentionsAnchors.ContextActionsAnchor,
  IntentionsAnchors.ContextActionsAnchorPosition);
```

It gets stuck around the string literal in the assertion. Sits at the top of [this loop](https://github.com/isagalaev/highlight.js/blob/master/src/highlight.js#L433-L440):

```cs
while (true) {
  top.terminators.lastIndex = index;
  match = top.terminators.exec(value);
  if (!match)
    break;
  count = processLexeme(value.substr(index, match.index - index), match[0]);
  index = match.index + count;
}
```

I'm at a bit of a loss as to why this is happening...

**Comment by citizenmatt** [comment]
Um. [Here's the regex](http://regexper.com/#%2F%5C%2F%5C%2F%5C%2F%7C%5C%2F%5C%2F%7C%5C%2F%5C*%7C%23%7C%40%22%7C'%7C%22%7C%5Cb(0%5BxX%5D%5Ba-fA-F0-9%5D%2B%7C(%5Cd%2B(%5C.%5Cd*)%3F%7C%5C.%5Cd%2B)(%5BeE%5D%5B-%2B%5D%3F%5Cd%2B)%3F)%7C(%5Ba-zA-Z%5D%5Cw*)*%5Cs*(class%7Cinterface%7Cstruct)%7C%5C.%3F(%5Cb(namespace)%5Cb)%5C.%3F%7C%5C.%3F(%5Cb(enum)%5Cb)%5C.%3F%7C%5C.%3F(%5Cb(new%7Creturn%7Cthrow%7Cawait)%5Cb)%5C.%3F%7C(%5Ba-zA-Z%5D%5Cw*(%3C%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F(%2C%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F)*%3E)%3F(%5B.%5D%5Ba-zA-Z%5D%5Cw*(%3C%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F(%2C%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F)*%3E)%3F)*%5Cs%2B)%2B%5Ba-zA-Z%5D%5Cw*(%3C%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F(%2C%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F)*%3E)%3F%5Cs*%5C(%7C(%5Ba-zA-Z%5D%5Cw*(%3C%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F(%2C%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F)*%3E)%3F(%5B.%5D%5Ba-zA-Z%5D%5Cw*(%3C%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F(%2C%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F)*%3E)%3F)*%5Cs%2B)%2B%5Ba-zA-Z%5D%5Cw*(%3C%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F(%2C%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F)*%3E)%3F%5Cs*(%5C%5B%5B%5E%5C%5D%5D%2B%5C%5D)%3F%5Cs*%7B%7C%5C%5B(%5Cs*(return%7Cassembly%7Cfield%7Cmethod%7Cevent%7Cparam)%5Cs*%3A)%3F%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F(%2C%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F)*%3E)%3F(%5B.%5D%5Ba-zA-Z%5D%5Cw*(%3C%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F(%2C%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F)*%3E)%3F)*%5Cs*(%5C(.%2B%3F%5C))%3F%5Cs*(%2C(%5Cs*(return%7Cassembly%7Cfield%7Cmethod%7Cevent%7Cparam)%5Cs*%3A)%3F%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F(%2C%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F)*%3E)%3F(%5B.%5D%5Ba-zA-Z%5D%5Cw*(%3C%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F(%2C%5Cs*%5Ba-zA-Z%5D%5Cw*(%3C%5B%5E%3E%5D%3E)%3F)*%3E)%3F)*%5Cs*(%5C(.%2B%3F%5C))%3F%5Cs*)*%5C%5D%7C%3A%3A%2Fgm) it's trying to evaluate. It's a bit, er, big.

**Comment by citizenmatt** [comment]
Actually, the size of the regexes isn't the problem. There were a couple of rogue quantifiers that were zero or many instead of zero or one. This has fixed my major near-infinite loops. Can you try the latest version of the PR and see how yours goes?

**Comment by damirarh** [comment]
The latest version performs much better. None of my code blocks caused any problems. The build time of my complete site is only around 5% slower than with version 8.3.0. It seems you have successfully fixed the issue.

**Comment by Sannis** [comment]
We can include few benchmarks into test suite to avoid such degradations in future, but I have no ideas how to perfectly automate this now.

**Comment by isagalaev** [comment]
Okay, if this is indeed fixed by #683 I'm going to close the issue as there's nothing more to do. Feel free to reopen it if I'm missing something!
